Code/Framework/AzFramework/AzFramework/Script/ScriptComponent.cpp

```diff
@@ -576,12 +581,18 @@ namespace AzFramework
     {
         DestroyEntityTable();
 
-        m_script = asset;
+        // Assign the updated asset on the next tick so we are not holding
+        // on to the old script
+        AZ::TickBus::QueueFunction(
+            [this, asset]()
+            {
+                m_script = asset;
 
-        if (LoadInContext())
-        {
-            CreateEntityTable();
-        }
+                if (LoadInContext())
+                {
+                    CreateEntityTable();
+                }
+            });
```

Comment:
I'm surprised this makes a difference because the reload notification should be invoked on the tick bus.

Should DestroyEntityTable also be deferred?

Using AZ::TickBus::QueueFunction can result in invalid memory access and crashes if the entity is destroyed before the queued function executes. It would be safer to store the new script asset, connect to the tick bus, then update the script and disconnect from the tick bus after it fires. This will also be more efficient in case the script asset changes multiple times before it's reloaded.

I also noticed that there was no bus for the script component. If there was a bus with an interface for Get/SetScript Then you could use QueueFunction and update the script using the bus. SetScript would need to reload the script. That would also allow assigning scripts dynamically, if there was a use case for that.

***At the very least, requesting that this changes to connecting and disconnecting from the tick bus to protect against invoking the queued function after the component is deactivated or deleted. Connect to the tick bus on asset reloaded, disconnect after its handled, disconnect on deactivation.

Reply:
Good catch, I'll modify to at least use the TickBus.  There is likely some other issue going on here because I think the OnAssetReloaded message shouldn't get sent until it's OK to reload the asset, however, there could be something going on where the old asset isn't getting freed up properly. Might dig into it further if I get time.